Add instrumentation using XRay

htuall opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):
- Experimental Feature

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):
Add instrumentation using XRay to parse xray_instr_map section generated by xray and map entry addresses to function names. Resolves #74249

### Details
Created parse_elf.cpp designated to parse xray_instr_map section generated by xray to then map entry addresses to function names. Currently not finished, needs further work, submitting as a checkpoint to track progress. Current problem: function address that I get from xray_instr_map does not match to the expected address of the function, which I also get from elf-binary. Possible reasons: 1) incorrect parsing of the section; 2) some relocations in elf-binary.

Resolves #74249

htuall committed:
working on parsing xray_instr_map section from elf-binary to map xray funcIds to function names

pamarcos commented:
Glad you're already working on this @htuall :tada:! Please next time ping me directly or even assign the PR to me so that I'm aware. Somehow this went off under my radar till now :eyes: 

Have you checked how the llvm-xray tool is doing this same thing? I assumed we could get the mapping [the same way](https://github.com/llvm/llvm-project/blob/c2979c58d49bf3c7dc892ed9fb49cdca389130ee/llvm/tools/llvm-xray/xray-extract.cpp#L78) and our part would be to find the proper function addresses based on non-mangled function names, which is what I think `readSymbolTable` is doing, right? We'll probably need to demangle those, or alternatively mangle the input symbols to be able to match them. The latter seems less work, since you only need to do that for those symbols we'll be interested in, rather than converting all symbols.

Also, I'm afraid I can't yet understand Russian comments that well even after @nikitamikhaylov's efforts to teach me the basics. So, please use English whenever possible so that I don't have to be going back and forth translating :pray:

htuall committed:
new logic for mapping functionIDs to their names + new experiment with custom handler for xray + started working on adding SYSTEM-statements to add/remove instrumentation(wip)

pamarcos commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -1329,6 +1339,20 @@ void InterpreterSystemQuery::loadOrUnloadPrimaryKeysImpl(bool load)
+    // Also we patch here I think(?)
+    // So we firstly map the function name to its ID(where do we store the mapping though)
+    // After that we need to do something like __xray_set_handler(functionID, handler, params);
+    // and __xray_patch(functionID)/__xray_unpatch(functionID)
+    // But does xray provide this kind of functionality?.. I am afraid it does not
```
Isn't this function exactly that? https://github.com/ClickHouse/llvm-project/blob/7e72cd6d7ae4cfabf1cc4a77eb2787ece492bb8b/compiler-rt/include/xray/xray_interface.h#L104-L106

```cpp
/// This patches a specific function id. See XRayPatchingStatus for possible
/// result values.
XRayPatchingStatus __xray_patch_function(int32_t FuncId)
```

htuall replied:
Thank you, you are totally right! Got a little mixed up for some reason, but now figured this out, so, indeed, no problems with setting a handler to a particular function.

pamarcos commented on src/Parsers/ParserSystemQuery.cpp:
```diff
@@ -733,6 +733,33 @@ bool ParserSystemQuery::parseImpl(IParser::Pos & pos, ASTPtr & node, Expected &
+                ASTPtr params;
+                if (!ParserStringLiteral{}.parse(pos, params, expected))
+                    return false;
+                res->parameters.emplace_back(replica_ast->as<ASTLiteral &>().value.safeGet<String>());
```
I got an error trying to compile this:

```
/home/ubuntu/ClickHouse/ClickHouse/src/Parsers/ParserSystemQuery.cpp:758:46: error: use of undeclared identifier 'replica_ast'
  758 |                 res->parameters.emplace_back(replica_ast->as<ASTLiteral &>().value.safeGet<String>());
      |                                              ^
1 error generated.
```

htuall replied:
Managed to fix this, but there are some problems with build due to llvm, I'll direct any remaining questions(and some updates) to Slack.

pamarcos reviewed (commented):
You're going in the right direction :)
I merged master against your branch and also removed the unnecessary binary artifacts that I think you added by mistake.

Please, make sure that ClickHouse is compiling correctly. Also some test even if very basic would be nice. Even just to ensure that the new system statement works.

htuall committed:
created XRayInstrumentationManager class + finished work on system statements using it + attempted to add system.instrumentation table(not sure if it's coorrect

htuall committed:
fix style + fix build

htuall committed:
fix build

htuall committed:
fix style

htuall committed:
attempt to fix build: does not find llvm files

htuall committed:
revert

htuall committed:
try to fix llvm includes

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Only build XRay-related instrumentation when XRay is enabled

Let's build with ENABLE_XRAY=1 only for the Debug build on AMD
for now. This is just to ensure on the CI that it builds
correctly.

There are some USE_XRAY missing in other files, but it is left
on purpose as an excercise to the reader :)

pamarcos committed:
Fix XRay includes

Commit:
only instrument if USE_XRAY=1

htuall committed:
fix unterminated conditional directive error + replaced ifdef with if(needs check) + includes to cpp from header

htuall committed:
fix return type

htuall committed:
add one more if USE_XRAY

htuall committed:
add include in cpp + if USE_XRAY

htuall committed:
add new system statements to show privileges test

htuall committed:
Merge branch 'master' into xray_experiments

htuall committed:
fix includes in XRayInstrumentationManager.h

htuall committed:
Merge branch 'xray_experiments' of github.com:htuall/ClickHouse into xray_experiments

htuall committed:
use correct xray function

htuall committed:
fix System/SystemStorageInstrumentation + fix style

htuall committed:
started working on system.instrumentation_profiling_log, in progress

pamarcos committed:
Fix XRay linking error

htuall committed:
define getColumnsDescription

htuall committed:
fix style

htuall committed:
fix new system statement: more intuitive order of parameters, parse handler_name and function_name as identifiers and not string literals, add possibility to pass different type of parameters for differnt hadlers and make them optional

htuall committed:
fix build errors

htuall committed:
fix parameters parsing

htuall committed:
working on InstrumentationProfilingLog

pamarcos commented on programs/server/config.xml:
```diff
@@ -1242,6 +1242,30 @@
+    </instrumentation>
+
+    <!-- XRay profiling contains rows with profiling information produced using XRay that we can export to Chrome trace format. -->
+    <xray_profiling>
+        <database>system</database>
+        <table>xray_profiling</table>
+        <flush_interval_milliseconds>7500</flush_interval_milliseconds>
+        <max_size_rows>1048576</max_size_rows>
+        <reserved_size_rows>8192</reserved_size_rows>
+        <buffer_size_rows_flush_threshold>524288</buffer_size_rows_flush_threshold>
+        <collect_interval_milliseconds>1000</collect_interval_milliseconds>
+        <flush_on_crash>false</flush_on_crash>
+    </xray_profiling>
```
I think this is not needed nor used anymore, right?

htuall replied:
I think I forget to rename this, it's supposed to be instrumentation_profiling_log, for which I think we discussed that we need this? Maybe you mean the one that is written for system.insertmentation? In this case I agree, I remember we discussed that for this kind of table it's not needed, I'll remove it in the next commit.

pamarcos replied:
Yes, you're right! I thought the other one was the `instrumentation_profiling_log` one :upside_down_face:

htuall committed:
added logic for stripping function names + throw exception for invalid handler or function + make parameters optional + do not need handler and parameters for remove

htuall committed:
fixed parsing but need to consider a better way to do this, rn seems like a duct tape

htuall committed:
never instrument instrumentWithXRay function

htuall committed:
added new handlers + pass and validate  parameters

htuall committed:
fix build

htuall committed:
fix style

htuall committed:
write to system.instrumentation_profiling_log in profile handler

htuall committed:
make system.instrumentation_profilling_log convertable to Chrome trace format

htuall committed:
fix style

htuall committed:
add shutdown

pamarcos commented on src/Interpreters/InstrumentationProfilingLog.cpp:
```diff
@@ -0,0 +1,68 @@
+    return ColumnsDescription
+    {
+        {"name", std::make_shared<DataTypeString>(), "Name of the instrumented function."},
+        {"cat", std::make_shared<DataTypeString>(), "Category of the instrumented function."},
+        {"ph", std::make_shared<DataTypeString>(), "Phase: entry or exit."},
+        {"pid", std::make_shared<DataTypeUInt64>(), "Process ID."},
```
I think we don't need ot do a 1:1 mapping of everything we need for the Chrome trace format. I think there's no need to save extra information if we can build those parameters based on some others:

- The PID is going to be the same
- Why do we need to save an entry and then an exit `phase`? Why don't we store the timestamp where something started + duration and then we build the two Chrome trace format rows from there? That way we avoid having 2 entries when we could have only 1. It's also easier to understand by a human in my opinion, cause the the whole event info is in a single row.
- We can use whatever in the `cat` part as far as I know, right?

pamarcos committed:
Revert "add shutdown"

This reverts commit f887e27bf80ffe5d2b40e51c0e8a2291fafc99e4.

pamarcos commented on src/Interpreters/InstrumentationProfilingLog.cpp:
```diff
@@ -0,0 +1,68 @@
+
+void InstrumentationProfilingLog::shutdown()
+{
+    stopFlushThread();
```
I understand you did this trying to fix the fast tests, but I don't think this is needed.
Check out the log from https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=77153&sha=f887e27bf80ffe5d2b40e51c0e8a2291fafc99e4&name_0=PR&name_1=Fast%20test: https://s3.amazonaws.com/clickhouse-test-reports/PRs/77153/f887e27bf80ffe5d2b40e51c0e8a2291fafc99e4//fast_test/clickhouse-server.log.zst

```
5. ./ci/tmp/build/./src/Storages/KeyDescription.cpp:163: DB::KeyDescription::getSortingKeyFromAST(std::shared_ptr<DB::IAST> const&, DB::ColumnsDescription const&, std::shared_ptr<DB::Context const>, std::optional<String> const&) @ 0x000000000e9f542e
6. ./ci/tmp/build/./src/Storages/KeyDescription.cpp:93: DB::KeyDescription::getKeyFromAST(std::shared_ptr<DB::IAST> const&, DB::ColumnsDescription const&, std::shared_ptr<DB::Context const>) @ 0x000000000e9f5edb
7. ./ci/tmp/build/./src/Storages/MergeTree/registerStorageMergeTree.cpp:642: DB::create(DB::StorageFactory::Arguments const&) @ 0x000000000f26325c
8. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x000000000ea4a886
9. ./ci/tmp/build/./src/Interpreters/InterpreterCreateQuery.cpp:1952: DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::unique_ptr<DB::DDLGuard, std::default_delete<DB::DDLGuard>>&, DB::LoadingStrictnessLevel) @ 0x000000000d6b0582
10. ./ci/tmp/build/./src/Interpreters/InterpreterCreateQuery.cpp:1750: DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x000000000d6a9610
11. ./ci/tmp/build/./src/Interpreters/InterpreterCreateQuery.cpp:2358: DB::InterpreterCreateQuery::execute() @ 0x000000000d6b5efc
12. ./ci/tmp/build/./src/Interpreters/SystemLog.cpp:729: DB::SystemLog<DB::InstrumentationProfilingLogElement>::prepareTable() @ 0x000000000d8f9550
13. ./ci/tmp/build/./src/Interpreters/SystemLog.cpp:562: DB::SystemLog<DB::InstrumentationProfilingLogElement>::savingThreadFunction() @ 0x000000000d8fa6f9
14. ./ci/tmp/build/./src/Common/SystemLogBase.cpp:283: ThreadFromGlobalPoolImpl<true, true>::ThreadFromGlobalPoolImpl<DB::SystemLogBase<DB::InstrumentationProfilingLogElement>::startup()::'lambda'()>(DB::SystemLogBase<DB::InstrumentationProfilingLogElement>::startup()::'lambda'()&&)::'lambda'()::operator()() @ 0x00000000093c9aab
15. ./contrib/llvm-project/libcxx/include/__functional/function.h:716: ? @ 0x0000000009380e40
16. ./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:117: void* std::__thread_proxy[abi:ne190107]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void (ThreadPoolImpl<std::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::thread>::ThreadFromThreadPool*>>(void*) @ 0x0000000009386a3b
17. ? @ 0x00007f879e3a6ac3
18. ? @ 0x00007f879e437a04
 (version 25.5.1.1)
2025.04.27 16:17:40.776961 [ 166049 ] {} <Debug> SystemLog (system.instrumentation_profiling_log): Creating new table system.instrumentation_profiling_log for InstrumentationProfilingLog
```

pamarcos committed:
Tune fields needed for system.instrumentation_profiling_log

pamarcos committed:
Clarify column name

htuall committed:
dispatch handler + new columns in system.instrumentation

htuall committed:
fix style

htuall committed:
fix build

htuall committed:
fix linker error

htuall committed:
small fixes

htuall committed:
small fix for profile

htuall committed:
small fix for system.instrumentation_profiling_log

htuall committed:
clean up code

htuall committed:
fix style

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos reviewed (commented):
Excellent work overall :raised_hands: 
I've added a few nitpick comments about the style and some other things that are really important about thread-safety.

We also need to think how to add tests for this. For simple methods such as `removeTemplateArgs` and `extractNearestNamespaceAndFunction` we can just add unit tests if there's no easy way to test it with a stateless or integration test.
For the rest, I'd try to think of some stateless tests that can check that this does what's supposed to do. They're simpler to write and check than integration tests. If we cannot think how to do it reliably with stateless tests, then we fallback to integration tests.

pamarcos committed:
Add XRay prefix to classes and tables

At first I thought not having the XRay was better, but I've
changed my mind. I prefer to make it very explicit. Also,
this way it's easier to filter what's related to XRay
instrumentation and what's not.

pamarcos commented:
Another thing I thought the other day: As of now, we're offering hooking only into the `function-enter` part, right? I thought it might be valuable and not so difficult to do something generic for all handlers (except for PROFILE) to be able to hook into both `function-enter` and `function-exit` using suffixes. e.g. `LOG_ENTER`/`LOG_EXIT` and `SLEEP_ENTER`/`SLEEP_EXIT`

What do you think?

htuall committed:
small changes about style + stop using call once + correctly set dispatch handler now

htuall committed:
updates

htuall committed:
Merge branch 'master' into xray_experiments

htuall committed:
accept handler names both in lower and upper case, use scope_exit for in_hook variable, use TSA_GUARDED_BY, fixed mutex usage -- less mutex locks and only when needed + one mutex for all three fields

htuall commented on src/Interpreters/XRayInstrumentationManager.h:
```diff
@@ -0,0 +1,90 @@
+    using HandlerTypeToIP = std::unordered_map<HandlerType, std::list<InstrumentedFunctionInfo>::iterator>;
+    InstrumentedFunctions getInstrumentedFunctions()
+    {
+        std::lock_guard lock(functions_to_instrument_mutex);
```
You are right, there was actually a couple unnecessary mutexes, including this one. I fixed it in the last commit.

htuall commented on src/Interpreters/XRayInstrumentationManager.cpp:
```diff
@@ -0,0 +1,516 @@
+    static thread_local bool in_hook = false;
+    if (in_hook) return;
+    in_hook = true;
+    std::shared_lock lock(shared_mutex);
```
I now only use one shared_mutex for every field that needs mutex at all, as they are always modified at the same time. For reading I use shared_lock.

pamarcos commented on src/Interpreters/XRayInstrumentationManager.h:
```diff
@@ -0,0 +1,90 @@
+    static std::unordered_map<int32_t, HandlerTypeToIP> functionIdToHandlers;
+
+    static inline std::mutex log_mutex;
+    static inline std::shared_mutex shared_mutex;
```
You can take a look and try if [UniqueLock](https://github.com/ClickHouse/ClickHouse/blob/31fedd140da5a2c05996a83a8c1aedfd46373f4c/src/Common/UniqueLock.h) is what you might need. It's a wrapper that is compatible with TSA annotations

htuall committed:
Merge branch 'master' into xray_experiments

htuall committed:
unit tests for removeTemplateArgs and extractNearestNamespaceAndFunction

htuall committed:
Merge branch 'xray_experiments' of github.com:htuall/ClickHouse into xray_experiments

pamarcos committed:
Use ClickHouse's SCOPE_EXIT instead of LLVM's one

pamarcos committed:
Avoid using static members when they're not needed

pamarcos committed:
Fix errors raised by TSA

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Bring back getInstrumentedFunctions

pamarcos committed:
Fix linter for unit test

pamarcos committed:
Fix removeTemplateArgs

pamarcos committed:
Fix extractNearestNamespaceAndFunction

pamarcos committed:
Fix using uppercase handler name

pamarcos committed:
Improve log

pamarcos committed:
Fix system instrument access

pamarcos committed:
Another fix for lowercase instrument

pamarcos committed:
Improve SYSTEM INSTRUMENT REMOVE to use instrumentation_point_id

This is much simpler, since we can use the unique ID for each
instrumentation point, instead of having to specify the function
name and the handler.

pamarcos committed:
Attempt to fix clang-tidy error in test

pamarcos committed:
Fix for real the clang-tidy error in unit test

pamarcos committed:
Rename instrumentation_point_id to just id

pamarcos committed:
Add INSTRUMENT REMOVE ALL option

pamarcos committed:
Add SYSTEM INSTRUMENT ADD/REMOVE test

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Add SYSTEM INSTRUMENT LOG and SLEEP test

pamarcos committed:
Fix style check

pamarcos committed:
Restore original name cause it's causing issues in sync

pamarcos committed:
Make the test faster

pamarcos committed:
Small improvements. Hold the mutex the minimum

pamarcos committed:
Add SYSTEM INSTRUMENT PROFILE test

pamarcos committed:
Consolidate the logger

pamarcos committed:
Consolidate 3 different maps with boost::multi_index

pamarcos committed:
Parse the XRay instrumentation map in parallel jobs

pamarcos committed:
Fix style check

pamarcos committed:
Enable XRay for amd_release and amd_binary

pamarcos committed:
Fix stripped function names not being unique

pamarcos committed:
Fix job size

pamarcos committed:
Add stack trace to LOG handler

pamarcos committed:
Add entry/exit option for LOG and SLEEP handlers

- Big refactor to avoid having different data structures that hold
  similar maps for instrumented points.

pamarcos committed:
Fix tests and issues found after refactor

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Fix USE_XRAY guard and linter

pamarcos committed:
Fix build for non-XRay flavor

pamarcos committed:
Add test for ENTRY/EXIT

pamarcos committed:
Make handler order deterministic

pamarcos committed:
Fix config with xray_instrumentation_profiling_log table

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Fix XRay instrumentation for tail exit

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Improve text in case function is not found

pamarcos committed:
Profile before sleep handler

pamarcos committed:
Enable XRay for all Linux flavors

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Remove XRay for sanitizers because clang_rt symbols clash

pamarcos commented:
It seems we cannot use XRay and any other sanitizer at the same time as per https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=77153&sha=23c6ed27502189ca2df435a42c653cb5ed18896c&name_0=PR

Even though I haven't found on the internet any evidence of them being incompatible, both sanitizers and XRay seem to use the same underlying allocator and runtime infrastructure from sanitizer_common, which is why you get "multiple definition" errors for symbols like __sanitizer::internal_allocator().

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Add documentation for the XRay's instrumentation feature

pamarcos committed:
Use Int32 everywhere for function_id to be consistent with XRay's signature

pamarcos committed:
Add words to dictionary

pamarcos committed:
Fix docs

pamarcos committed:
Enable XRay for ARM

pamarcos committed:
Use more LowCardinality when it makes sense

pamarcos committed:
Add more documentation

pamarcos committed:
Don't use XRay along with sanitizer for ARM since they're not compatible

pamarcos committed:
Fix style check
